### PR TITLE
Fix confusion with column reordering on arrays

### LINKED ISSUE #427: Column Reordering Bug?
I'm just trying out the column reordering at https://github.com/vaadin/vaadin-grid/commit/d252b3f94f10cdce8f693b3ebfefb7f5f30c5237 and I'm getting a weird behaviour where switching the last two columns seems to do nothing but change the headers sometimes.

See: 
reorder-bug

Relevant code snippet is: 
```html
<vaadin-grid column-reordering-allowed frozen-columns="1">
        <table>
          <colgroup>
            <col>
            <col>
            <col>
            <col>
          </colgroup>
          <thead>
          <tr>
            <th>Frozen column</th>
            <th>First Name</th>
            <th>Last Name</th>
            <th>Gender</th>
          </tr>
          </thead>
          <tbody>
          <tr>
            <td>1</td>
            <td>Alice</td>
            <td>Williamson</td>
            <td>Male</td>
          </tr>
          <tr>
            <td>2</td>
            <td>Florence</td>
            <td>Lewis</td>
            <td>Male</td>
          </tr>
          <tr>
            <td>3</td>
            <td>Trevor</td>
            <td>Ramirez</td>
            <td>Female</td>
          </tr>
          <tr>
            <td>4</td>
            <td>Herbert</td>
            <td>Collins</td>
            <td>Genderqueer</td>
          </tr>
          <tr>
            <td>5</td>
            <td>Roy</td>
            <td>Spencer</td>
            <td>Male</td>
          </tr>
          </tbody>
        </table>
      </vaadin-grid>
```

There's nothing fancy there.

### PR DESCRIPTION
**Pull request by tehapo**
Fixes #427

Column reordering does not work as expected when not using objects as data. This misleading behavior is described in #427.

This PR does some refactoring, improves related tests, but most importantly prevents the misleading partial reordering (by calling `refreshItems()`) and documents the limitations of this feature explicitly.

---
This change is [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-grid/432)

### CODE REVIEW

**Comment by manolo** [comment]
I miss some check in order not to enable the feature if the data does not matches the requirements. 
I think we can reuse the setColumns and check whether the name attribute is used and enable/disable the flag conveniently.

---

Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[demo/columns.html, line 283 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/432#-KPr3X1EPFgmAUnrlnkt:-KPr3X1FnMlbNjH2ApA0:b-onrm7r) ([raw file](https://github.com/vaadin/vaadin-grid/blob/1c6b8c48de31dbe077f937b56f6a718ca9054e87/demo/columns.html#L283)):*
> ```HTML
>       as a data source and bind their properties to the column configuration
>       by using the <code>name</code> attribute.
>     </p>
> ```

I would rephrase this:

`Column reordering is only supported when the data source is based on Objects bound to the column configuration using the <code>name</code> attribute.`

---

*[docs/vaadin-grid-columns.adoc, line 172 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/432#-KPr4wzf5786El7RrKUt:-KPr4wzg9vu3HmBJH-_c:b-hxexdz) ([raw file](https://github.com/vaadin/vaadin-grid/blob/1c6b8c48de31dbe077f937b56f6a718ca9054e87/docs/vaadin-grid-columns.adoc#L172)):*
> ```AsciiDoc
> 
> [NOTE]
> When using the column reordering feature, you must use objects as values and bind them to the column configuration using the `name` attribute.
> ```

Copy the same text from above.

---

*[java/src/main/java/com/vaadin/elements/grid/GridElement.java, line 142 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/432#-KPr-DhMFx77CXNnGric:-KPr-DhN-0AcD7IWdL5B:b-750vq4) ([raw file](https://github.com/vaadin/vaadin-grid/blob/1c6b8c48de31dbe077f937b56f6a718ca9054e87/java/src/main/java/com/vaadin/elements/grid/GridElement.java#L142)):*
> ```Java
>         return order;
>     }
>     
> ```

remove spaces

---

*[java/src/main/java/com/vaadin/elements/grid/GridElement.java, line 145 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/432#-KPr-HtnyPuhv9mf6pLC:-KPr-HtnyPuhv9mf6pLD:b-88f5my) ([raw file](https://github.com/vaadin/vaadin-grid/blob/1c6b8c48de31dbe077f937b56f6a718ca9054e87/java/src/main/java/com/vaadin/elements/grid/GridElement.java#L145)):*
> ```Java
>     private void updateOrder() {
>         cols.setLength(0); // Clear.
>         
> ```

remove spacels

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/432#-:-KPr51aCeqYGQwBTxGbO:b-z8k69n)*

**Comment by tehapo** [comment]
Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[demo/columns.html, line 283 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/432#-KPr3X1EPFgmAUnrlnkt:-KPrWagFSll9IbqJ7ZMB:b6qr9zc) ([raw file](https://github.com/vaadin/vaadin-grid/blob/1c6b8c48de31dbe077f937b56f6a718ca9054e87/demo/columns.html#L283)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary>
> I would rephrase this:
> 
> `Column reordering is only supported when the data source is based on Objects bound to the column configuration using the <code>name</code> attribute.`
> 
</details>
Thanks, that sounds better than the original wording. :+1: 

---

*[docs/vaadin-grid-columns.adoc, line 172 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/432#-KPr4wzf5786El7RrKUt:-KPrWdYm_Z9s4jEeqBf1:b-896fix) ([raw file](https://github.com/vaadin/vaadin-grid/blob/1c6b8c48de31dbe077f937b56f6a718ca9054e87/docs/vaadin-grid-columns.adoc#L172)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary>
> Copy the same text from above.
</details>
Done.

---

*[java/src/main/java/com/vaadin/elements/grid/GridElement.java, line 142 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/432#-KPr-DhMFx77CXNnGric:-KPrWf4PhX_PT7aroilX:b-896fix) ([raw file](https://github.com/vaadin/vaadin-grid/blob/1c6b8c48de31dbe077f937b56f6a718ca9054e87/java/src/main/java/com/vaadin/elements/grid/GridElement.java#L142)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary>
> remove spaces
</details>
Done.

---

*[java/src/main/java/com/vaadin/elements/grid/GridElement.java, line 145 \[r1\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/432#-KPr-HtnyPuhv9mf6pLC:-KPrWfLPaj6Z7b8qSUl7:b-896fix) ([raw file](https://github.com/vaadin/vaadin-grid/blob/1c6b8c48de31dbe077f937b56f6a718ca9054e87/java/src/main/java/com/vaadin/elements/grid/GridElement.java#L145)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary>
> remove spacels
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/432)*

**Comment by manolo** [comment]
Reviewed 2 of 6 files at r1, 4 of 4 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/432)*

**Comment by manolo** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/432#-:-KPv5Dqd1QOXEZ_spedl:bnfp4nl)*

**Comment by manolo** [comment]
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-grid/432)*

File: docs/vaadin-grid-columns.adoc [hunk 1]
```diff
@@ -168,6 +168,11 @@ endif::web[]
 The end users can reorder non-frozen columns by dragging the header cells.
 By default, column reordering is disabled, but can be enabled by setting the [propertyname]#columnReorderingAllowed# property to true.
 
+[NOTE]
+Column reordering is only supported when the data source is based on objects bound to the column configuration using the `name` attribute.
```
Comment (tomivirkki) [inline docs/vaadin-grid-columns.adoc]: It should work as well with array datasources as long as the source returns correctly ordered data after column reorder
